### ISSUE TITLE
Versioned::get_versionnumber_by_stage does not support fluent

### ISSUE THREAD

**Issue by blueo** [tractorcow-farm/silverstripe-fluent#583]
Don't know if this belongs on Fluent or Versioned but anyhow: the `get_versionnumber_by_stage` method on Versioned (line 2435) uses a simple (optimised) query to get a version number for an object:
```
   // get version as performance-optimized SQL query (gets called for each record in the sitetree)
        $version = DB::prepared_query(
            "SELECT \"Version\" FROM \"$stageTable\" WHERE \"ID\" = ?",
            [$id]
        )->value();
```
where stageTable is the table for the base class. This clearly does not take into account versions in different locales as the base table will contain the last written version number regardless of locale. The result being that you may get a 'later' version number than what exists in your current locale. 

the knock on effect for versioning is fairly wide-ranging as this is used in quite a few of the other versioned methods eg `stageDiffer` and `isLiveVersion`. 

We've been testing something like the following query instead:
```
SELECT CASE
	WHEN SiteTree_Localised_Versions_Localised_en_NZ.ID IS NOT NULL THEN SiteTree_Localised_Versions_Localised_en_NZ.Version
ELSE BT.Version END AS Version
FROM SiteTree BT
LEFT JOIN SiteTree_Localised_Versions AS SiteTree_Localised_Versions_Localised_en_NZ ON BT.ID = SiteTree_Localised_Versions_Localised_en_NZ.RecordID AND SiteTree_Localised_Versions_Localised_en_NZ.Locale = 'en_NZ'
WHERE BT.ID = ?
ORDER BY Version DESC LIMIT 1;
```

but this requires using injector to subclass Versioned - so it doesn't feel like a great way to solve it? Any guidance on contributing this back would be appreciated.

**Comment by tractorcow** [comment]
Yes this is a problem, since each locale will have its own version it was saved in. I think you have the general idea of the problem and how to solve it; It's just getting it to let us override the versioned method.

**Comment by blueo** [comment]
all good then - will aim to add it as an extension point on versioned

**Comment by tractorcow** [comment]
Updating versioned to use the ORM instead of raw SQL could get us the same effect without a new extension point.
